viadot/sources/cloud_for_customers.py

```diff
@@ -5,56 +5,169 @@
+url = "https://my341115.crm.ondemand.com/sap/c4c/odata/ana_businessanalytics_analytics.svc/$metadata?entityset=RPZ1F90A299C830A29C6659D3QueryResult"
+
+
+def column_mapper(url=url):
+    credentials = local_config.get("CLOUD_FOR_CUSTOMERS")
+    access = credentials.get("Prod")
+    auth = (access["username"], access["password"])
+    response = requests.get(url, auth=auth)
+    column_mapping = {}
+    for sentence in response.text.split("/>"):
+        result = re.search(r'(?<=Name=")([^"]+).+(sap:label=")([^"]+)+', sentence)
+        if result:
+            key = result.groups(0)[0]
+            val = result.groups(0)[2]
+            column_mapping[key] = val
+    return column_mapping
+
+
+def change_to_meta_url(url):
+    meta_url = ""
+    start = url.split(".svc")[0]
+    ending = url.split("/")[-1]
+    e = ending.split("?")[0]
+    meta_url = start + ".svc/$metadata?entityset=" + e
+    return meta_url
+
+
+def response_to_entity_list(dirty_json, url):
+    metadata_url = change_to_meta_url(url)
+    column_maper_dict = column_mapper(metadata_url)
+    entity_list = []
+    for element in dirty_json["d"]["results"]:
+        new_entity = {}
+        for key, object_of_interest in element.items():
+            if (
+                key != "__metadata"
+                and key != "Photo"
+                and key != ""
+                and key != "Picture"
+            ):
+                # skip the column which contain nested structure
+                if "{" not in str(object_of_interest):
+                    new_key = column_maper_dict.get(key)
+                    if new_key:
+                        new_entity[new_key] = object_of_interest
+                    else:
+                        new_entity[key] = object_of_interest
+        entity_list.append(new_entity)
+    return entity_list
+
+
+class APIError(Exception):
+    pass
 
-    Args:
-        url (str, optional): The url to the API. Defaults to None.
-        endpoint (str, optional): The endpoint of the API. Defaults to None.
-        params (Dict[str, Any]): The query parameters like filter by creation date time. Defaults to json format.
-    """
 
+class CloudForCustomers(Source):
     def __init__(
         self,
         *args,
+        direct_url: str = None,
         url: str = None,
         endpoint: str = None,
         params: Dict[str, Any] = {},
-        **kwargs
+        **kwargs,
     ):
+        """
+        Fetches data from Cloud for Customer.
+
+        Args:
+            direct_url (str, optional): The url to the API in case of prepared report. Defaults to None.
+            url (str, optional): The url to the API. Defaults to None.
+            endpoint (str, optional): The endpoint of the API. Defaults to None.
+            params (Dict[str, Any]): The query parameters like filter by creation date time. Defaults to json format.
+        """
         super().__init__(*args, **kwargs)
         credentials = local_config.get("CLOUD_FOR_CUSTOMERS")
```

Comment:
This breaks the Liskov Substitution Principle, since user should be able to pass `credentials` to any subclass of `Source`. See my other comment about this above.